File: resources/c++_and_stl/ch3-Problem_Solving_Methods/ch3_06_UVa108.cpp
/* Maximum Sum */

#include <iostream>
using namespace std;

#define REP(i, a, b) 	for (int i = int(a); i <= int(b); i++)

int main() { // O(n^4) DP solution
  int i, j, k, l, n, arr[101][101], maxSubRect, subRect;

  scanf("%d", &n);
  REP (i, 0, n - 1)
    REP (j, 0, n - 1) {
      scanf("%d", &arr[i][j]);
      if (i > 0) // 2nd row onwards
        arr[i][j] += arr[i - 1][j]; // add from top
      if (j > 0) // 2nd col onwards
        arr[i][j] += arr[i][j - 1]; // add form left
      if (i > 0 && j > 0)
        arr[i][j] -= arr[i - 1][j - 1]; // to avoid double count
    }

  maxSubRect = -127*100*100; // lowest possible value for this problem
  REP (i, 0, n - 1) REP (j, 0, n - 1)
    REP (k, i, n - 1) REP (l, j, n - 1) {
      subRect = arr[k][l];
      if (i > 0) subRect -= arr[i - 1][l];
      if (j > 0) subRect -= arr[k][j - 1];
      if (i > 0 && j > 0) subRect += arr[i - 1][j - 1];
      maxSubRect = max(maxSubRect, subRect);
    }

  printf("%d\n", maxSubRect);

  return 0;
}



/*
// brute force, TLE
#include <algorithm>
#include <iostream>
using namespace std;

int main() {
  int n, arr[101][101], maxSubRect, subRect;

  scanf("%d", &n);
  for (int i = 0; i < n; i++) for (int j = 0; j < n; j++)
    scanf("%d", &arr[i][j]);

  maxSubRect = -127*100*100; // lowest possible value for this problem
  for (int i = 0; i < n; i++) for (int j = 0; j < n; j++) // start coord
    for  (int k = i; k < n; k++) for (int l = j; l < n; l++) { // end coord
      subRect = 0; // sum items in this sub-rectangle
      for (int a = i; a <= k; a++) for (int b = j; b <= l; b++)
        subRect += arr[a][b];
      maxSubRect = max(maxSubRect, subRect); // keep largest so far
    }

  printf("%d\n", maxSubRect);
}

*/



/*
#include <iostream>
using namespace std;

#define REP(i, a, b) 	for (int i = int(a); i <= int(b); i++)

int main() { // O(n^3) DP solution
  int n, unCounted = 1, currentMax = 0, possibleMax, matrix[101][101];

  scanf("%d", &n);
  REP (i, 0, n - 1)
    REP (j, 0, n - 1) {
      scanf("%d", &matrix[i][j]);
      if (j > 0) matrix[i][j] += matrix[i][j - 1]; // pre-processing
    }

  REP (left, 0, n - 1)
    REP (right, left, n - 1) {
      possibleMax = 0;
      REP (row, 0, n - 1) {
        if (left > 0)
          possibleMax += matrix[row][right] - matrix[row][left - 1];
        else
          possibleMax += matrix[row][right];
        
        if (possibleMax < 0) possibleMax = 0;

        if (unCounted || possibleMax > currentMax) {
          currentMax = possibleMax;
          unCounted = 0;
        }
      }
    }

  printf("%d\n", currentMax);
} // return 0;
*/

File: resources/c++_and_stl/ch3-Problem_Solving_Methods/ch3_06_UVa108.java
import java.util.*;

class Main { /* Maximum Sum */
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int i, j, left, right, row, n, unCounted = 1, currentMax = 0, possibleMax;
    int[][] matrix = new int[101][101];

    n = sc.nextInt();
    for (i = 0; i < n; i++)
      for (j = 0; j < n; j++) {
        matrix[i][j] = sc.nextInt();
        if (j > 0) matrix[i][j] += matrix[i][j - 1]; // pre-processing
      }

    // O(n^3) greedy solution (Kadane's algorithm)
    for (left = 0; left < n; left++)
      for (right = left; right < n; right++) {
        possibleMax = 0;
        for (row = 0; row < n; row++) {
          if (left > 0)
            possibleMax += matrix[row][right] - matrix[row][left - 1];
          else
            possibleMax += matrix[row][right];
        
          // crucial part, immediately reset if running sum is negative
          if (possibleMax < 0) possibleMax = 0;

          if (unCounted == 1 || possibleMax > currentMax) {
            currentMax = possibleMax;
            unCounted = 0;
          }
        }
      }

    System.out.printf("%d\n", currentMax);
  }
}
